contrib/win32/win32compat/tncon.c

```diff
@@ -155,632 +157,644 @@ ReadConsoleForTermEmul(HANDLE hInput, char *destin, int destinlen)
-						if(!modKey)
-							NetWriteString2(pParams->Socket, (char *)INSERT_KEY, 4, 0);
-						else {
-							/* ^[[2;m~ */
-							char *p = "\033[2;";
-							strcpy_s(tmp_buf, sizeof(tmp_buf), p);
-							size_t index = strlen(p);
-							tmp_buf[index++] = modKey + '0';
-							tmp_buf[index] = '~';
-
-							NetWriteString2(pParams->Socket, tmp_buf, index+1, 0);
-						}
-						break;
-					case VK_DELETE:
-						if(!modKey)
-							NetWriteString2(pParams->Socket, (char *)REMOVE_KEY, 4, 0);
-						else {
-							/* ^[[3;m~ */
-							char *p = "\033[3;";
-							strcpy_s(tmp_buf, sizeof(tmp_buf), p);
-							size_t index = strlen(p);
-							tmp_buf[index++] = modKey + '0';
-							tmp_buf[index] = '~';
-
-							NetWriteString2(pParams->Socket, tmp_buf, index+1, 0);
-						}
-						break;
-					case VK_PRIOR: /* page up */
-						if (!modKey)
-							NetWriteString2(pParams->Socket, (char *)PREV_KEY, 4, 0);
-						else {
-							/* ^[[5;m~ */
-							char *p = "\033[5;";
-							strcpy_s(tmp_buf, sizeof(tmp_buf), p);
-							size_t index = strlen(p);
-							tmp_buf[index++] = modKey + '0';
-							tmp_buf[index] = '~';
-
-							NetWriteString2(pParams->Socket, tmp_buf, index+1, 0);
-						}
-						break;
-					case VK_NEXT: /* page down */
-						if(!modKey)
-							NetWriteString2(pParams->Socket, (char *)NEXT_KEY, 4, 0);
-						else {
-							/* ^[[6;m~  */
-							char *p = "\033[6;";
-							strcpy_s(tmp_buf, sizeof(tmp_buf), p);
-							size_t index = strlen(p);
-							tmp_buf[index++] = modKey + '0';
-							tmp_buf[index] = '~';
-
-							NetWriteString2(pParams->Socket, tmp_buf, index+1, 0);
-						}
-						break;
-					case VK_BACK:
-						NetWriteString2(pParams->Socket, (char *)BACKSPACE_KEY, 1, 0);
-						break;
-					case VK_TAB:
-						if (dwControlKeyState == SHIFT_PRESSED)
-							NetWriteString2(pParams->Socket, (char *)SHIFT_TAB_KEY, 3, 0);
-						else
+		ReadConsoleInputW(hInput, inputRecordArray, inputRecordArraySize, &dwInput);
+
+		for (int i=0; i < dwInput; i++) {
+			INPUT_RECORD inputRecord = inputRecordArray[i];
+
+			switch (inputRecord.EventType) {
+			case WINDOW_BUFFER_SIZE_EVENT:
+				queue_terminal_window_change_event();
+				break;
+
+			case FOCUS_EVENT:
+				/* FALLTHROUGH */
+			case MENU_EVENT:
+				break;
+
+			case KEY_EVENT:
+				bCapsOn = (inputRecord.Event.KeyEvent.dwControlKeyState & CAPSLOCK_ON);
+				bShift = (inputRecord.Event.KeyEvent.dwControlKeyState & SHIFT_PRESSED);
+				dwControlKeyState = inputRecord.Event.KeyEvent.dwControlKeyState &
+					~(CAPSLOCK_ON | ENHANCED_KEY | NUMLOCK_ON | SCROLLLOCK_ON);
+
+				/* ignore the AltGr flags*/
+				if ((dwControlKeyState & dwAltGrFlags) == dwAltGrFlags)
+					dwControlKeyState = dwControlKeyState & ~dwAltGrFlags;
+
+				modKey = GetModifierKey(dwControlKeyState);
+				if (inputRecord.Event.KeyEvent.bKeyDown) {
+					int n = WideCharToMultiByte(
+						CP_UTF8,
+						0,
+						&(inputRecord.Event.KeyEvent.uChar.UnicodeChar),
```

Comment:
You will need to be careful here. Sometimes, you will get two input records that each contain half of a UTF-16 sequence -- this will mean that you get two failed conversions in a row

Reply:
If you send an emoji, U+1F574, you will get two KEY EVENTS:

```
0xD83D
0xDD74
```

If you convert them one at a time, you will get two `0xFFFD`, `0xFFFD` because individually they are not valid.

Reply:
fixed